helpers/analytics_helpers.rb

```diff
@@ -0,0 +1,42 @@
+module AnalyticsHelpers
+  def urlize(string)
+    string.downcase.gsub(" ", "_").gsub("-", "_")
+  end
+
+  def build_event(data, attributes = [], output = {})
+    data.sort_by { |k, _v| k["name"] }.each do |item|
+      if item["value"]
+        case item["value"]
+        when String
+          output[item["name"]] = item["value"]
+        when Array
+          output[item["name"]] = build_event(item["value"], attributes)
+        end
+      else
+        attribute = attributes.select { |x| x["name"] == item["name"] }
+        output[item["name"]] = if attribute && attribute.first
+                                 attribute.first["type"]
+                               else
+                                 item["value"]
+                               end
+      end
+    end
+    output
+  end
+
+  def to_html(hash, html = "")
+    html += "<ul class='govuk-list indented-list'>"
+    hash.each do |key, value|
+      case value
+      when String
+        html += "<li><a href='/analytics/attribute_#{urlize(key)}.html'>#{key}</a>: #{value}</li>"
+      when Hash
+        html += "<li>#{key}: "
+        html += to_html(value)
+        html += "</li>"
```

Comment:
Non-blocking: you might be able to use `content_tag` to reduce some of the lines here, although from reading the docs I'm not certain if it returns text, you might have to change the whole approach, in which case don't worry.

```Ruby
html += content_tag(:li, to_html(value))
```

Reply:
Thanks @andysellick - I've cleaned this up a bit.